File: python-bindings/tests/game_orchestration_conformance.py
"""Run real concurrent game checks and the public full-scan composition."""

import tempfile
from pathlib import Path


def observe_orchestration(fixture: dict) -> dict:
    """Authenticate report assembly before normalizing concurrent completion order."""
    import classic_scangame

    with tempfile.TemporaryDirectory(
        prefix="classic-orchestration-conformance-"
    ) as temporary:
        root = Path(temporary)
        for name, content in fixture["files"].items():
            if name != "sentinel.txt":
                raise ValueError("unsupported orchestration fixture path")
            (root / name).write_bytes(content.encode())
        config = classic_scangame.GameScanConfig(
            root,
            fixture["xseAcronym"],
            fixture["crashgenName"],
            fixture["gameName"],
            log_catch_errors=["error"],
        )
        if (
            config.game_path != root
            or config.xse_acronym != fixture["xseAcronym"]
            or config.game_name != fixture["gameName"]
        ):
            raise ValueError(
                "orchestrator configuration lost public constructor fields"
            )
        orchestrator = classic_scangame.GameScanOrchestrator(config)

        def game_result(value):
            """Verify raw report order, then canonicalize inherently unordered job completion."""
            checks = [
                {"name": row.name, "output": row.output} for row in value.check_results
            ]
            matches = value.report == "".join(row["output"] for row in checks)
            return {
                "checkResults": sorted(checks, key=lambda row: row["name"]),
                "reportMatchesChecks": matches,
                "configIssueCount": len(value.config_issues),
                "errors": sorted(value.errors),
            }

        def mod_result(value):
            """Read every public mod-scan summary field."""
            return {
                "report": value.report,
                "unpackedIssueCount": value.unpacked_issue_count,
                "archivedIssueCount": value.archived_issue_count,
                "errors": sorted(value.errors),
            }

        game = game_result(orchestrator.run_game_checks())
        mods = mod_result(orchestrator.run_mod_scans())
        full_game, full_mods = orchestrator.run_full_scan()
        if game_result(full_game) != game or mod_result(full_mods) != mods:
            raise ValueError("full scan lost one of its native component results")
        return {
            "game": game,
            "mods": mods,
            "files": [
                {
                    "path": path.relative_to(root).as_posix(),
                    "content": path.read_bytes().decode(),
                }
                for path in sorted(root.rglob("*"))
                if path.is_file()
            ],
        }

File: node-bindings/classic-node/__test__/game_orchestration_conformance.ts
import { mkdtemp, writeFile, readdir, readFile, rm } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import * as classic from "../index.js";

/** Observe native game and mod stages, retaining nonfatal configuration errors. */
export async function observeOrchestration(fixture: any): Promise<Record<string, unknown>> {
  const root=await mkdtemp(join(tmpdir(),"classic-orchestration-conformance-"));
  try {
    for (const [name,content] of Object.entries(fixture.files)){if(name!=="sentinel.txt")throw new Error("unsupported orchestration fixture path");await writeFile(join(root,name),String(content));}
    const config={gamePath:root,xseAcronym:fixture.xseAcronym,crashgenName:fixture.crashgenName,gameName:fixture.gameName,xseScriptfiles:{},isVr:false,gameVersion:"Original",wryeWarnings:{},logCatchErrors:["error"],logExcludeFiles:[],logExcludeErrors:[],gameTarget:"Fallout4"};
    const result=await classic.runGameChecks(config);const mods=await classic.runModScans(config);
    // JoinSet completion order varies; authenticate the native report before canonicalizing it.
    const reportMatchesChecks=result.report===result.checkResults.map(check=>check.output).join("");
    const game={checkResults:[...result.checkResults].sort((a,b)=>a.name.localeCompare(b.name)),reportMatchesChecks,configIssueCount:result.configIssues.length,errors:[...result.errors].sort()};
    const files=[];for(const name of (await readdir(root)).sort())files.push({path:name,content:await readFile(join(root,name),"utf8")});
    return {game,mods:{report:mods.report,unpackedIssueCount:mods.unpackedIssueCount,archivedIssueCount:mods.archivedIssueCount,errors:[...mods.errors].sort()},files};
  } finally {await rm(root,{recursive:true,force:true});}
}
